Now using the metadata action for tags & labels

GoliathLabs opened the pull request:
We now use the metadata action for tags & labels of our images (https://github.com/docker/metadata-action#usage, https://docs.github.com/en/actions/publishing-packages/publishing-docker-images)

Please have a look if I did everything right

Closes #68

GoliathLabs commented:
The output of the build action looks promising so far 
![image](https://user-images.githubusercontent.com/8057646/150851384-a24f4ca5-c1b0-4490-bb4a-cfaf8078223f.png)

DasSkelett commented:
You need to look at the "Docker inspect" output to see whether it actually worked. Looks like it didn't:
```
"Labels": null
```

DasSkelett commented:
```
$${{ steps.meta.output.labels }}
```
The `$$` looks a bit suspicious, is it intentional?

DasSkelett commented on .github/workflows/build.yml:
```diff
@@ -33,11 +35,7 @@ jobs:
-            org.opencontainers.image.source=${{ github.event.repository.clone_url }}
-            org.opencontainers.image.revision=${{ github.sha }}
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: $${{ steps.meta.output.labels }}
```
The `$$` looks a bit suspicious, is it intentional?

GoliathLabs replied:
You're right. Gonna fix this

GoliathLabs commented:
Hmm. It still ain't happy

DasSkelett commented:
Yeah, now it appears to be `null`.
I already tried the action when we added the opencontainers labels initially, and discovered that it doesn't work. That's why I went with setting them manually, which was simple, reliable and did everything we needed.

But maybe you can get it to work after playing around more.

DasSkelett commented on .github/workflows/build.yml:
```diff
@@ -33,11 +35,7 @@ jobs:
-            org.opencontainers.image.source=${{ github.event.repository.clone_url }}
-            org.opencontainers.image.revision=${{ github.sha }}
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.output.labels }}
```
```suggestion
          labels: ${{ steps.meta.outputs.labels }}
```

GoliathLabs commented:
> Yeah, now it appears to be `null`. I already tried the action when we added the opencontainers labels initially, and discovered that it doesn't work. That's why I went with setting them manually, which was simple, reliable and did everything we needed.
> 
> But maybe you can get it to work after playing around more.

I've corrected the typo there and the labels are now available

GoliathLabs reviewed (commented):
Therefore I'd go with a LGTM 😄

DasSkelett commented:
Did you compare the labels before/after to make sure we didn't loose any, or otherwise break something?
Before:
```
"org.opencontainers.image.authors": "Freifunk München",
"org.opencontainers.image.created": "2022-01-24 15:04:26+00:00",
"org.opencontainers.image.revision": "4da0a8a36dc30616349796b35b38d1b8a73a96fc",
"org.opencontainers.image.source": "https://github.com/freifunkMUC/wg-access-server.git"
```
Now:
```
"org.opencontainers.image.created": "2022-01-24T20:41:28.645Z",
"org.opencontainers.image.description": "An all-in-one WireGuard VPN solution with a web ui for connecting devices",
"org.opencontainers.image.licenses": "MIT",
"org.opencontainers.image.revision": "540c6867e10c9f945d6391f7d94873d2d5e4e91a",
"org.opencontainers.image.source": "https://github.com/freifunkMUC/wg-access-server",
"org.opencontainers.image.title": "wg-access-server",
"org.opencontainers.image.url": "https://github.com/freifunkMUC/wg-access-server",
"org.opencontainers.image.version": "pr-70"
```
`authors` is gone, for example, it would make sense to add it back (the action supports specifying additional ones, I think).

GoliathLabs commented:
Yeah, you're right. Yet I can't find a reference to `authors` in their usage documentation. Otherwise the authors can be seen by looking at the `image.source` or `image.url` tags I think

DasSkelett commented:
If I'm understanding [this section](https://github.com/docker/metadata-action#customizing) correctly, then it should be possible to specify additional custom labels.

DasSkelett commented:
This will make images built from commits to master be tagged as `master` and no longer `latest`, right?
And `latest` will be tagged when new releases are created?

If so, this change is probably important enough to be mentioned in the pull request body, so those who review the PR (and anyone else who looks at it in the future) doesn't have to figure it out on their own (or worse, only realize it after it has been merged).

GoliathLabs commented:
@DasSkelett if i read this section correctly it should tag new images automatically as latest https://github.com/docker/metadata-action#latest-tag

DasSkelett commented:
> @DasSkelett if i read this section correctly it should tag new images automatically as latest [docker/metadata-action#latest-tag](https://github.com/docker/metadata-action#latest-tag)

So that section says:
> `latest` tag is handled through the `flavor` input. It will be generated by default (auto mode) for:
>   - `type=ref,event=tag`
>   - `type=semver,pattern=...`
>   - `type=match,pattern=...`

The docs for `type=ref` say:
> This type handles Git ref (or reference) for the following events:
>   - branch ; eg. refs/heads/master
>   - **tag ; eg. refs/tags/v1.0.0**
>   - pr ; eg. refs/pull/318/merge
> | Event           | Ref                           | Output                        |
> |-----------------|-------------------------------|-------------------------------|
> | `pull_request`  | `refs/pull/2/merge`           | `pr-2`                        |
> | `push`          | `refs/heads/master`           | `master`                      |
> | `push`          | `refs/heads/my/branch`        | `my-branch`                   |
> | `push tag`      | `refs/tags/v1.2.3`            | `v1.2.3`                      |
> | `push tag`      | `refs/tags/v2.0.8-beta.67`    | `v2.0.8-beta.67`              |

This sounds to me like it would only trigger when git tags are pushed/created.

`type=semver` section says:
> **Will be used on a push tag event** and requires a valid semver Git tag, but you can also use a custom value through value attribute.

`type=match` section says:
> Can create a regular expression for matching Git tag with a pattern and capturing group. **Will be used on a push tag event** but you can also use a custom value through value attribute.

That README is _really_ confusingly written, but looking at it one by one it sounds like that `latest` image tag is only attached for git tag builds.
Which I'm not necessarily against, that's at least how most projects do it. It just means that we have to do regular releases now, but @awlx planned to do that anyway in #68.

GoliathLabs commented:
Yes, you are right. But then do we get a master build instead? Or how does the action handle regular pushes then? Since we'll definitely get a push on latest somehow at some point (at least when we push to a tag), I'd say we can merge the PR. The README of the metadata action is really a nightmare though. Maybe after we have our experience with the gh action, we can create a clarifying pull request.

DasSkelett commented:
Yeah I think they'll be tagged `master`. Let's find out!
(Also squashed your commits so the commit history of the master branch is a bit more tidy)

GoliathLabs commented:
Perfect, thank you very much. I've also created a new issue in the action's repo. Maybe there'll be some clarification :)
